### PR TITLE
Masterbar - always show site items + add prevSelectedSiteId to calypso state

### PR DESCRIPTION
**Pull request by Addison-Stavlo**
Related to # p1721229953008389-slack-C06DN6QQVAQ

## Proposed Changes

Allows site specific items to remain in the masterbar at global levels where selectedSite is cleared for various reasons.

* Adds a `prevSelectedSiteId` to the calypso redux store. We do not need an action for this, we just set dispatches to it when calls for `setSelectedSiteId` or `setAllSitesSelected` are called. `prevSelectedSiteId` may be null initially, but we prevent setting null values on changes to always have context of the most previously selected site.
    * Updates and adds unit tests related to these actions, reducer, and selectors. 
    * Updated 2 `useDispatch` imports in stepper files that were using this selector with the useDispatch directly from react-redux. The types from react-redux do not seem to support thunks, so we import it from calypso/state instead which types the redux dispatch to support thunks.
* Updates the masterbar to always try to use a site. The priority goes from `currentlySelectedSite` > `previouslySelectedSite` > `primarySite`. This effects the site Menu, +new menu, and checkout cart.
    *  We may want to rethink or separate conditions for when we show which items for which sites. But this seems like a good starting point to try.

BEFORE
prev-site-before

AFTER
prev-site-after

## Why are these changes being made?

 Context from p1721229953008389-slack-C06DN6QQVAQ
* We want the global masterbar to be consistent while navigating throughout global pages. The domains, themes, plugins, etc. at global levels are tangled with the selectedSiteId - deselecting it on load to accommodate the global context and usage of these pages.
* Adding a prevSelectedSiteId and applying these changes to state seems like the safest way to ensure continuity when navigating around calypso global levels, and these priorities for siteId in the masterbar allow for it to always have context for a site if one exists.

## Testing Instructions

* Go to calypso /sites and select a site to overview.
* verify the site menu, +new,  etc, links continue work as expected.
* close the overview tab, verify the site menu does not disappear.
* verify the site menu is still present (previously would disappear at this point) and that, +new, etc. links continue to work as expected.
* click into the domains tab on the global sidebar. Verify the site specific items are still in the global sidebar. verify the masterbar and this domains page work as expected.
* do the above for global plugins and themes pages.
* smoke test around calypso. verify things work as expected and we haven't created and weird edge cases.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Review by davemart-in — APPROVED** [review]
This is awesome. Exactly what I was hoping for! Everything worked perfectly in my testing.

File: client/landing/stepper/declarative-flow/new-hosted-site-flow-user-included.ts [hunk 1]
```diff
@@ -4,14 +4,13 @@ import { NEW_HOSTED_SITE_FLOW_USER_INCLUDED } from '@automattic/onboarding';
...
 import { useEffect, useLayoutEffect } from 'react';
-import { useDispatch as reduxUseDispatch } from 'react-redux';
 import { recordFreeHostingTrialStarted } from 'calypso/lib/analytics/ad-tracking/ad-track-trial-start';
 import {
 	setSignupCompleteSlug,
 	persistSignupDestination,
 	setSignupCompleteFlowName,
 } from 'calypso/signup/storageUtils';
-import { useSelector } from 'calypso/state';
+import { useDispatch as reduxUseDispatch, useSelector } from 'calypso/state';
```
Comment (Addison-Stavlo) [inline client/landing/stepper/declarative-flow/new-hosted-site-flow-user-included.ts:13]: There were type errors here trying to use dispatch straight from react-redux, as they aren't typed to support thunks like they are when imported from calypso/state.

File: client/landing/stepper/declarative-flow/new-hosted-site-flow.ts [hunk 2]
```diff
@@ -3,14 +3,13 @@ import { NEW_HOSTED_SITE_FLOW } from '@automattic/onboarding';
...
 import { useEffect, useLayoutEffect } from 'react';
-import { useDispatch as reduxUseDispatch } from 'react-redux';
 import { recordFreeHostingTrialStarted } from 'calypso/lib/analytics/ad-tracking/ad-track-trial-start';
 import {
 	setSignupCompleteSlug,
 	persistSignupDestination,
 	setSignupCompleteFlowName,
 } from 'calypso/signup/storageUtils';
-import { useSelector } from 'calypso/state';
+import { useDispatch as reduxUseDispatch, useSelector } from 'calypso/state';
```
Comment (Addison-Stavlo) [inline client/landing/stepper/declarative-flow/new-hosted-site-flow.ts:12]: There were type errors here trying to use dispatch straight from react-redux, as they aren't typed to support thunks like they are when imported from calypso/state.

File: client/landing/stepper/declarative-flow/new-hosted-site-flow-user-included.ts (hunk 1 above)
Comment (Addison-Stavlo) [reply client/landing/stepper/declarative-flow/new-hosted-site-flow-user-included.ts:13]: note it looks like there are some places in stepper already using this version of useDispatch, ex:
https://github.com/Automattic/wp-calypso/blob/64dc80d4412606544a6f0e32febf52f618c7d251/client/landing/stepper/declarative-flow/site-setup-flow.ts#L11
https://github.com/Automattic/wp-calypso/blob/64dc80d4412606544a6f0e32febf52f618c7d251/client/landing/stepper/declarative-flow/internals/steps-repository/design-setup/unified-design-picker.tsx#L51
etc.

Comment (Addison-Stavlo) [reply client/landing/stepper/declarative-flow/new-hosted-site-flow-user-included.ts:13]: It seems that calypso/state useDispatch is [literally just the same useDispatch from react-redux](https://github.com/Automattic/wp-calypso/blob/64dc80d4412606544a6f0e32febf52f618c7d251/client/state/index.ts#L57), but[ typed to support thunks.](https://github.com/Automattic/wp-calypso/blob/64dc80d4412606544a6f0e32febf52f618c7d251/client/state/types.ts#L32) so this seems safe.

**Review by DustyReagan — APPROVED** [review]
I followed the testing instructions, and tried to break this by clicking around. It's working great!

**Review by okmttdhr — APPROVED** [review]
Beautifully done :) ✅ Thank you, @Addison-Stavlo! 
📝 This also addresses that showing the primary site when no site is selected: https://github.com/Automattic/dotcom-forge/issues/8324#issuecomment-2233254139

I'm shipping this change.

**Pull request merged**